src/main.cpp

```diff
@@ -1363,32 +1363,23 @@ void vcRenderSceneUI(vcState *pProgramState, const ImVec2 &windowPos, const ImVe
     {
       // Compass
       {
-        udDouble3 cameraDirection = udDirectionFromYPR(pProgramState->camera.eulerRotation);
-        udDouble3 northDirection = udDouble3::create(0, 1, 0);
+        udDouble3 cameraDirection = vcGIS_HeadingPitchToQuaternion(pProgramState->gis, pProgramState->camera.position, pProgramState->camera.headingPitch).apply({ 0, 1, 0 });
 
         udDouble3 up = vcGIS_GetWorldLocalUp(pProgramState->gis, pProgramState->camera.position);
-
-        if (pProgramState->gis.isProjected)
-        {
-          udDouble3 currentLatLong = udGeoZone_CartesianToLatLong(pProgramState->gis.zone, pProgramState->camera.position);
-          currentLatLong.x = udClamp(currentLatLong.x, -90.0, 89.9);
-          northDirection = udNormalize(udGeoZone_LatLongToCartesian(pProgramState->gis.zone, udDouble3::create(currentLatLong.x + 0.1, currentLatLong.y, currentLatLong.z)) - pProgramState->camera.position);
-        }
-
-        udDouble3 east = udCross(northDirection, up);
-        udDouble3 northFlat = udCross(up, east);
+        udDouble3 northDir = vcGIS_GetWorldLocalNorth(pProgramState->gis, pProgramState->camera.position);
 
         udDouble3 camRight = udCross(cameraDirection, up);
         udDouble3 camFlat = udCross(up, camRight);
 
-        double angle = udATan2(camFlat.x * northFlat.y - camFlat.y * northFlat.x, camFlat.x * northFlat.x + camFlat.y * northFlat.y);
+        double angle = udATan2(camFlat.x * northDir.y - camFlat.y * northDir.x, camFlat.x * northDir.x + camFlat.y * northDir.y);
```

Comment:
This is fine, but it is not immediately obvious what is happening here; we are using a combination of the perpendicular dot product `camFlat.x * northDir.y - camFlat.y * northDir.x` and the dot product `camFlat.x * northDir.x + camFlat.y * northDir.y` to find the angle between two vectors. This doesn't need to change but is another example where we can be a little more verbose to give clues to the reader as to what is happening at a deeper level.